### PR TITLE
[GEP-26] Fix cloudprovider secret deployment to not renew WorkloadIdentity token on shoot reconciliation

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area security quality ipcei
/kind bug
/label ipcei/workload-identity

**What this PR does / why we need it**:
Fix cloudprovider secret deployment to not renew WorkloadIdentity token on shoot reconciliation.
Currently, all already existing annotations, labels and data fields are removed on each shoot reconciliation, most of these changes triggers the workload identity token requestor controller to renew the token.

**Which issue(s) this PR fixes**:
Follow up on https://github.com/gardener/gardener/pull/10239
Part of https://github.com/gardener/gardener/issues/9586
Alternative of https://github.com/gardener/gardener/pull/10371

**Special notes for your reviewer**:
cc @rfranzke @vpnachev 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/utils/workloadidentity/secret.go
Comment (rfranzke) [inline pkg/utils/workloadidentity/secret.go]: Maybe just `WithProviderConfig`?

File: pkg/utils/workloadidentity/secret.go
Comment (rfranzke) [inline pkg/utils/workloadidentity/secret.go]: `WithContextObject`?

File: pkg/utils/workloadidentity/secret.go
Comment (rfranzke) [inline pkg/utils/workloadidentity/secret.go]: Just `For`?

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/secrets.go:493]:
```suggestion
```
Let's not tamper with the annotations if not needed :)

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/secrets.go:496]: Use `metav1.SetMetaDataLabel` instead :) Then we don't tamper with the labels either

**Review by rfranzke — COMMENTED** [review]
Very nice @dimityrmirchev !

File: pkg/apis/security/v1alpha1/constants/constants.go
Comment (vpnachev) [inline pkg/apis/security/v1alpha1/constants/constants.go]: #10313 introduced this [constant](https://github.com/gardener/gardener/pull/10313/files#diff-578591f0d06c51b53b9d3760b243d573bbc3bc15c2a6d5119c4ed744ff87a96bR19-R21), I would expect github to find a merge conflict but I do not know why it does not. To avoid issue (duplicated definition of constant), please rebase the PR to the latest master and drop the duplication.

File: pkg/apis/security/v1alpha1/constants/constants.go
Comment (vpnachev) [inline pkg/apis/security/v1alpha1/constants/constants.go]:
```suggestion
	// DataKeyToken is the data key of a secret whose value contains a workload identity token.
	DataKeyToken = "token"
	// DataKeyConfig is the data key of a secret whose value contains a workload identity provider configuration.
	DataKeyConfig = "config"
```

`DataKeyToken` and `DataKeyConfig` belong to the same group of constants, thus you can remove the empty line separating them.

File: pkg/utils/workloadidentity/secret_test.go
Comment (vpnachev) [inline pkg/utils/workloadidentity/secret_test.go]: The test seems to be interacting only with exposed APIs of the package, therefore you can add `_test` suffix to the package name 
```suggestion
package workloadidentity_test
```

File: pkg/utils/workloadidentity/workloadidentity_suite_test.go
Comment (vpnachev) [inline pkg/utils/workloadidentity/workloadidentity_suite_test.go]: What is the reason to not use `workloadidentity_test`?

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (vpnachev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: Actually, it is on purpose. Otherwise when switching from workload identity to static infra credentials label like `security.gardener.cloud/purpose=workload-identity-token-requestor` will not be removed and the workload identity token requestor controller will keep reconciling this secret. Same is valid for the annotations.

File: pkg/utils/workloadidentity/secret_test.go
Comment (vpnachev) [inline pkg/utils/workloadidentity/secret_test.go]:
```suggestion
		Expect(err.Error()).To(And(
			ContainSubstring("workload identity name is not set"),
			ContainSubstring("workload identity namespace is not set"),
			ContainSubstring("workload identity provider type is not set"),
		))
```

Not necessary a change request, but you might like it.

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: Doesn't it make sense to remove these well-known labels/annotations (and vice-versa)? This way, we would still allow extensions to add additional labels/annotations similar to how it's done for other resources.

Comment (vpnachev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: If the extensions are managing the annotations/labels via admission webhooks, then there is no change for them as they will intercept the update request and ensure they are not removed.

Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: Yes, but if controllers add these labels/annotations, then they get removed in the next shoot reconciliation. I think it's easy to support this scenario since the labels/annotations are well-known.

Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: To be honest even the previous implementation was overwriting labels and annotations so we are not doing anything different here. https://github.com/gardener/gardener/blob/706bf6564134b34a89d02c4994803743c277df3a/pkg/gardenlet/operation/botanist/secrets.go#L472-L477

> I think it's easy to support this scenario since the labels/annotations are well-known.

Maybe we can do it but I do we see a use case for it right now. Provider extensions use mutating webhooks to inject additional data into the cloudprovider secret.

The workload identity secret also overwrites any labels and annotations that could be set by other webhooks/controllers. https://github.com/gardener/gardener/blob/8f04d573e1c9f83d8723d0752df8d723a1f81606/pkg/utils/workloadidentity/secret.go#L163-L166

File: pkg/utils/workloadidentity/secret.go
Comment (dimityrmirchev) [reply pkg/utils/workloadidentity/secret.go]: Although the length of the function name is not ideal I think it is still easier to grasp for a user of the package what this function will do if we leave the `WorkloadIdentity` in the function (option) name. When importing the package there can be some duplication in the form of `workloadidentity.WithWorkloadIdentityProviderConfig`, but I think that this is acceptable. This is also my opinion for `WithWorkloadIdentityContextObject` and `ForWorkloadIdentity` functions.

File: pkg/utils/workloadidentity/workloadidentity_suite_test.go
Comment (dimityrmirchev) [reply pkg/utils/workloadidentity/workloadidentity_suite_test.go]: It seems weird that this function is in a `_test` package but then tests actually use internals of that package. See https://github.com/gardener/gardener/blob/8f04d573e1c9f83d8723d0752df8d723a1f81606/pkg/utils/workloadidentity/jwt_test.go.

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: I know, but since the code is now touched anyways, why not use the chance to improve it? I don't have a concrete use-case in mind, but I'm also not asking for something overwhelmingly complex. This "cloud provider secret copying"-code is very old and we rarely had to touch/adjust, but we paid attention in newer code to not blindly overwrite all labels/annotations. Hence, I think it's fair to consider this as "cleaning up a bit of technical debts" and streamlining things. 

> The workload identity secret also overwrites any labels and annotations that could be set by other webhooks/controllers.



Good point, I think I missed this. For the `ServiceAccount` access secrets, we don't do it, see https://github.com/gardener/gardener/blob/9a6cc7d06c9415ba6ed5ab07b0fbc854e5761924/pkg/utils/gardener/shoot.go#L323-L448

File: pkg/utils/workloadidentity/secret.go
Comment (rfranzke) [reply pkg/utils/workloadidentity/secret.go]: I don't agree - almost every usage of these functions will be in other packages, so it always looks like your example `workloadidentity.<function>`. The package name already scopes semantically what these functions are related to. It's even an official code style best practice in Go to avoid stuttering (it's just doesn't kick in here because of the `With` or `For` prefixes).

This is an example complaint (when no prefixes are used):

```
file.go:4:6: type name will be used as file.FileName by other packages, and that stutters; consider calling this Name
file.go:6:6: type name will be used as file.FilePath by other packages, and that stutters; consider calling this Path
```

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: My concerns are that the following code is weird

```go
secret, _ := workloadidentity.NewSecret(
	name,
	namespace,
	workloadidentity.WithLabels(map[string]string{"foo": "bar"}),
)
_ = secret.Reconcile(ctx, client)

secret, _ = workloadidentity.NewSecret(
	name,
	namespace,
	workloadidentity.WithLabels(map[string]string{"another-foo": "bar"}),
)
_ = secret.Reconcile(ctx, client)
```

In this case the label `"foo": "bar"` will be preserved on the secret. One would have to remember to clean this separately.

Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: This is a valid concern and indeed the price for this practice. However, this also applies to the vast majority of the other code in this repository, so why do something different here?

Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: > However, this also applies to the vast majority of the other code in this repository, so why do something different here?

Correct me if I am wrong but most of the resources are applied via MRs and that will not preserve any labels or annotations.

The `AccessSecret` seem to respect setting labels/annotations only during create. After that the `Reconcile` function only cares about its known labels/annotations. A user does not seem to have direct access to change these via the `Reconcile` func. https://github.com/gardener/gardener/blob/9a6cc7d06c9415ba6ed5ab07b0fbc854e5761924/pkg/utils/gardener/shoot.go#L385

The managed resource secrets also overwrite labels and annotations https://github.com/gardener/gardener/blob/9a6cc7d06c9415ba6ed5ab07b0fbc854e5761924/pkg/utils/managedresources/builder/secrets.go#L109-L116

Comment (vpnachev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: Having in mind that the cloudprovider secret is in the seed cluster where the access is very restricted by design, it should not be expected a human user to manually set label or annotation (I am even not aware of a use case that requires such on this particular secret). If a controllers sets additional label or annotation, I see no issue if gardenlet does not preserve them because if the additional labels/annotations are set via:
- webhook they are actually never unset
- controller, it should observe the update of the resource, reconcile it and re-add the annotations/labels.

File: pkg/utils/workloadidentity/workloadidentity_suite_test.go
Comment (vpnachev) [reply pkg/utils/workloadidentity/workloadidentity_suite_test.go]: Is there a requirement that all `*_test.go` files must be of the same package?

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: Folks, I will resolve this conversation now. I understand that we may have different perspectives on this, and I respect your viewpoints. However, I believe that even minor improvements, especially when they come at a low cost, can lead to a better and more consistent codebase in the long run.
In this light, I have to express some disappointment. The change I proposed is indeed minor and IMO a step towards improving our codebase now that you touch this logic anyways. I've invested considerable time into this discussion, drawing from past experiences and learnings, and it's discouraging to see these suggestions dismissed without substantial counterarguments.
Let's proceed with the current approach for now, but I hope that in the future, we can have more constructive discussions about such improvements. I believe that every opportunity for improvement, no matter how small, is valuable.

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: > However, I believe that even minor improvements, especially when they come at a low cost, can lead to a better and more consistent codebase in the long run.

I never argued with that.

> it's discouraging to see these suggestions dismissed without substantial counterarguments.
> I hope that in the future, we can have more constructive discussions about such improvements

I cannot agree that counterarguments were not provided or that the discussion is not constructive. Please see https://github.com/gardener/gardener/pull/10383#discussion_r1741594768 and https://github.com/gardener/gardener/pull/10383#discussion_r1741672229. I think I gave an argument why this behaviour is unexpected for users of the said package. You call it improvement, but I actually see it as something that can cause bugs if labels/annotations are not cleaned up. I also pointed out how `AccessSecret` is not intuitive because you can set `secret.Labels`, but that only works during secret creation (this I did not test but the code looks like it is the case) and that MR secrets overwrite annotations and labels.

Let's not merge this if it is going to result in disappointment or bad feelings. I would appreciate if someone else from the core team expresses their opinion on the concrete matter. I will then proceed with whatever is the consensus.

/hold

File: pkg/utils/workloadidentity/workloadidentity_suite_test.go
Comment (dimityrmirchev) [reply pkg/utils/workloadidentity/workloadidentity_suite_test.go]: No, I think not. An interesting approach that I found in the go std lib https://github.com/golang/go/blob/master/src/bufio/export_test.go. A file that is part of the package itself is used in order to export functions for testing. Then this file has a `_test` suffix so that it is not included in the output binary.

I can apply this "export" approach or revert the change that I made with regards to the package naming. WDYT?

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: As said above, I am fine with continuing this PR without this adaptation. We seem to have different opinions of preserving labels/annotations, and that's perfectly fine for now, especially since it's technically unrelated to the PR as such. Hence, leave it aside for now. We might or might not consider doing this as a separate follow-up, but in the scope of this PR, it's not necessary to prolong this discussion any further or involve more people and their time.

File: pkg/utils/workloadidentity/workloadidentity_suite_test.go
Comment (vpnachev) [reply pkg/utils/workloadidentity/workloadidentity_suite_test.go]: Applying the test export pattern seems nice, if it is not much work, it will be nice contribution. Otherwise I am fine with just reverting to `_test` package name.

Comment (dimityrmirchev) [reply pkg/utils/workloadidentity/workloadidentity_suite_test.go]: Done with https://github.com/gardener/gardener/pull/10383/commits/14dd01874511562769c7676bcc440e14da4e081c

File: pkg/gardenlet/operation/botanist/secrets.go
Comment (dimityrmirchev) [reply pkg/gardenlet/operation/botanist/secrets.go:496]: /unhold

**Pull request merged**
